ext/serializer.c

```diff
@@ -828,6 +856,33 @@ static void _serialize_meta(zval *el, ddtrace_span_data *span) {
         add_assoc_str(meta, "_dd.span_links", buf.s);
     }
 
+    zend_array *peer_service_sources_zv = ddtrace_spandata_property_peer_service_sources(span);
+    if (peer_service_set) { // peer.service is already set by the user, honor it
+        add_assoc_str(meta, "_dd.peer.service.source", zend_string_init(ZEND_STRL("peer.service"), 0));
+    } else if (zend_hash_num_elements(peer_service_sources_zv) > 0) {
+        zval *tag;
+        ZEND_HASH_FOREACH_VAL(peer_service_sources_zv, tag) {
+            if (Z_TYPE_P(tag) == IS_STRING) {
+                // Use the first tag that is found in the span, if any
+                zval * value = zend_hash_find(Z_ARRVAL_P(meta), Z_STR_P(tag));
+                if (value) {
+                    add_assoc_str(meta, "_dd.peer.service.source", zend_string_copy(Z_STR_P(tag)));
+
+                    zend_string * substitute_value = dd_get_mapped_peer_service(Z_STR_P(value));
+                    if (substitute_value) {
+                        add_assoc_str(meta, "peer.service", substitute_value);
+                        add_assoc_str(meta, "peer.service.remapped_from", zend_string_copy(Z_STR_P(value)));
```

Comment:
This logic is missing in the explicitly set peer.service case. I assume it should be remapping then, too?

Reply:
Indeed, it was missing. I changed the logic a bit (refactored) to make use of the `MAP` configuration type